client/session.go

```diff
@@ -2005,28 +2010,35 @@ func (s *session) streamBlocksReattemptFromPeers(
 	// getting done because new attempts are blocked on existing attempts completing
 	// and existing attempts are trying to enqueue into a full reattempt channel
 	enqueue := enqueueCh.enqueueDelayed()
-	go s.streamBlocksReattemptFromPeersEnqueue(blocks, enqueue)
+	go s.streamBlocksReattemptFromPeersEnqueue(blocks, attemptErr, enqueue)
 }
 
 func (s *session) streamBlocksReattemptFromPeersEnqueue(
 	blocks []blockMetadata,
+	attemptErr error,
 	enqueueFn func([]*blocksMetadata),
 ) {
 	for i := range blocks {
 		// Reconstruct peers metadata for reattempt
 		reattemptBlocksMetadata :=
 			make([]*blocksMetadata, len(blocks[i].reattempt.peersMetadata))
 		for j := range reattemptBlocksMetadata {
+			reattempt := blocks[i].reattempt
+			// Copy the errors for every peer so they don't shard the same error
+			// slice and therefore are not subject to race conditions when the
+			// error slice is modified
+			reattempt.errs = append([]error{attemptErr}, reattempt.errs...)
```

Comment:
Perhaps `append(append([]error(nil), reattempt.errs...), attemptErr)` for in order errors?